// webdriverio/build/commands/browser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.waitUntil = exports.url = exports.uploadFile = exports.touchAction = exports.throttle = exports.switchWindow = exports.setWindowSize = exports.setTimeout = exports.setCookies = exports.saveScreenshot = exports.saveRecordingScreen = exports.reloadSession = exports.react$ = exports.react$$ = exports.pause = exports.newWindow = exports.mock = exports.keys = exports.getWindowSize = exports.getPuppeteer = exports.getCookies = exports.executeAsync = exports.execute = exports.deleteCookies = exports.debug = exports.custom$ = exports.custom$$ = exports.call = exports.$ = exports.$$ = void 0;
exports.$$ = require('./browser/$$').default;
exports.$ = require('./browser/$').default;
exports.call = require('./browser/call').default;
exports.custom$$ = require('./browser/custom$$').default;
exports.custom$ = require('./browser/custom$').default;
exports.debug = require('./browser/debug').default;
exports.deleteCookies = require('./browser/deleteCookies').default;
exports.execute = require('./browser/execute').default;
exports.executeAsync = require('./browser/executeAsync').default;
exports.getCookies = require('./browser/getCookies').default;
exports.getPuppeteer = require('./browser/getPuppeteer').default;
exports.getWindowSize = require('./browser/getWindowSize').default;
exports.keys = require('./browser/keys').default;
exports.mock = require('./browser/mock').default;
exports.newWindow = require('./browser/newWindow').default;
exports.pause = require('./browser/pause').default;
exports.react$$ = require('./browser/react$$').default;
exports.react$ = require('./browser/react$').default;
exports.reloadSession = require('./browser/reloadSession').default;
exports.saveRecordingScreen = require('./browser/saveRecordingScreen').default;
exports.saveScreenshot = require('./browser/saveScreenshot').default;
exports.setCookies = require('./browser/setCookies').default;
exports.setTimeout = require('./browser/setTimeout').default;
exports.setWindowSize = require('./browser/setWindowSize').default;
exports.switchWindow = require('./browser/switchWindow').default;
exports.throttle = require('./browser/throttle').default;
exports.touchAction = require('./browser/touchAction').default;
exports.uploadFile = require('./browser/uploadFile').default;
exports.url = require('./browser/url').default;
exports.waitUntil = require('./browser/waitUntil').default;
